featran-jmh/src/test/scala/com/spotify/featran/jmh/HotEncoderTransformersBenchmark.scala

```diff
@@ -0,0 +1,234 @@
+import org.openjdk.jmh.annotations._
+
+@BenchmarkMode(Array(Mode.AverageTime))
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Thread)
+class HotEncoderTransformersBenchmark {
+
+  //
+  // N HOT - optional
+  //
+  @Benchmark
+  def nHotOptionalEncoder: Unit = nHotOptional(arbitraryListString)
+
+  @Benchmark
+  def nHotOptionalEncoder1000Values: Unit =
+    nHotOptional(arbitraryListString1000Values)
+
+  @Benchmark
+  def nHotOptionalEncoder100000Values: Unit =
+    nHotOptional(arbitraryListString100000Values)
+
+  @Benchmark
+  def nHotOptionalEncoder100000ValuesWR: Unit =
+    nHotOptional(arbitraryListString100000ValuesWR)
+
+  //
+  // N HOT - required
+  //
+  @Benchmark
+  def nHotRequiredEncoder: Unit = nHotRequired(arbitraryListString)
+
+  @Benchmark
+  def nHotRequiredEncoder1000Values: Unit =
+    nHotRequired(arbitraryListString1000Values)
+
+  @Benchmark
+  def nHotRequiredEncoder100000Values: Unit =
+    nHotRequired(arbitraryListString100000Values)
+
+  @Benchmark
+  def nHotRequiredEncoder100000ValuesWR: Unit =
+    nHotRequired(arbitraryListString100000ValuesWR)
+
+  //
+  // N HOT - required: transform optional to list
+  //
+  @Benchmark
+  def nHotRequiredWithListMatch: Unit =
+    nHotRequiredWithListMatchPrepare(arbitraryListOptional)
+
+  @Benchmark
+  def nHotRequiredWithListMatch1000Values: Unit =
+    nHotRequiredWithListMatchPrepare(arbitraryListOptional1000Values)
+
+  @Benchmark
+  def nHotRequiredWithListMatch100000Values: Unit =
+    nHotRequiredWithListMatchPrepare(arbitraryListOptional100000Values)
+
+  @Benchmark
+  def nHotRequiredWithListMatch100000ValuesWR: Unit =
+    nHotRequiredWithListMatchPrepare(arbitraryListOptional100000ValuesWR)
+
+  //
+  // ONE HOT - optional
+  //
+  @Benchmark
+  def oneHotOptionalEncoder: Unit = oneHotOptional(arbitraryString)
+
+  @Benchmark
+  def oneHotOptionalEncoder1000Values: Unit =
+    oneHotOptional(arbitraryString1000Values)
+
+  @Benchmark
+  def oneHotOptionalEncoder100000Values: Unit =
+    oneHotOptional(arbitraryString100000Values)
+
+  @Benchmark
+  def oneHotOptionalEncoder100000ValuesWR: Unit =
+    oneHotOptional(arbitraryString100000ValuesWR)
+
+  //
+  // ONE HOT - required
+  //
+  @Benchmark
+  def oneHotRequiredEncoder: Unit = oneHotRequired(arbitraryString)
+
+  @Benchmark
+  def oneHotRequiredEncoder1000Values: Unit =
+    oneHotRequired(arbitraryString1000Values)
+
+  @Benchmark
+  def oneHotRequiredEncoder100000Values: Unit =
+    oneHotRequired(arbitraryString100000Values)
+
+  @Benchmark
+  def oneHotRequiredEncoder100000ValuesWR: Unit =
+    oneHotRequired(arbitraryString100000ValuesWR)
+
+  def nHotOptional(
+      input: List[List[String]]): FeatureExtractor[List, List[String]] = {
+    val featureSpec = FeatureSpec
+      .of[List[String]]
+      .optional {
+        case List("") => None
+        case i        => Option(i)
+      }(NHotEncoder("n_hot"))
+    featureSpec.extract(input)
```

Comment:
I'd move `val featureSpec` out of these methods so they're not measured in the benchmark.
IIRC `.extract()` returns a `FeatureExtractor`, but doesn't actually perform any computation until you call `.featureValues`?

Reply:
That makes a lot more sense - I'm seeing significant differences in benchmarks with more/less values when calling `.featureValues`.